wire-ios-data-model/Source/Model/UserClient/UserClient+Patches.swift

```diff
@@ -17,24 +17,56 @@
 //
 
 import Foundation
+import WireSystem
 
 extension UserClient {
 
     /// Migrate client sessions from using the client identifier only as session identifier
     /// to new client sessions  useing user identifier + client identifier as session identifier.
     /// These have less chances of collision.
+
     static func migrateAllSessionsClientIdentifiersV2(in moc: NSManagedObjectContext) {
+        WireLogger.proteus.info("migrating all session ids to v2")
+
         let request = UserClient.sortedFetchRequest()
 
-        guard
-            let allClients = moc.fetchOrAssert(request: request) as? [UserClient],
-            let keyStore = moc.zm_cryptKeyStore
-        else {
-            // no client? no migration needed
+        guard let allClients = moc.fetchOrAssert(request: request) as? [UserClient] else {
+            WireLogger.proteus.info("migrating all session ids to v2: no clients to migrate")
             return
         }
 
-        // TODO: [John] use flag here
+        // Important:
+        // If we're migrating from Cryptobox to Core Crypto, it's important that we
+        // migrate the session ids before moving the session files into Core Crypto.
+        //
+        // However, we would only have an instance of `ProteusService` (backed by Core
+        // Crypto) and not an instance of the `UserClientKeyStore` (backed by Cryptobox).
+        //
+        // Migration of the session ids relies on the keystore, so if we don't have an
+        // instance if the keystore already, we create one temporarily for this migration
+        // assuming that we'll perform the migration from Cryptobox to Core Crypto afterwards.
+
+        let keyStore: UserClientKeysStore
+
+        if let existingKeyStore = moc.zm_cryptKeyStore {
+            WireLogger.proteus.info("migrating all session ids to v2: using existing keystore")
+            keyStore = existingKeyStore
+        } else {
+            WireLogger.proteus.info("migrating all session ids to v2: creating temp keystore")
+
+            guard
+                let accountDirectory = moc.accountDirectoryURL,
```

Comment:
is this else case just a safe case but can't really happen?

Reply:
> how come the moc has this setup and not a UserClientKeysStore as setupUserKeyStore is called just after that
I don't quite understand the first question.

> is this else case just a safe case but can't really happen?
I understood that when `DeveloperFlag.proteusViaCoreCrypto` is enabled we wouldn't create the keystore, to avoid having both the keystore and the proteus service. I don't think we made that change yet but it is another PR.

So this else case is both being safe, but also anticipating that we won't have the keystore when moving to core crypto.